book/src/rfcs/RFC0007-Idempotent-Synchronized.md

```diff
@@ -0,0 +1,226 @@
+new scheme, the index `i` can be used to directly access the slab and check for
+a memoized value. Computing whether a query has changed since `r` begins with
+verifying dependencies. If verification fails, and there is a memoized value available
+whose `changed_at` is earlier than `r`, then we can compute the
+value for the current index (as described earlier) and see whether that value
+is the same as the memoized value (i.e., is the value backdated). If so, 
+we can return that the query has not changed.
+
+### Evicting for LRU
+
+The LRU system (for now, at least) can continue to work roughly as it does
+today, but with a few tweaks. For one thing, the LRU system today stores a
+`Arc<Node>` values that directly reference slots. We would instead store the key
+index values (i.e., the `usize`) that index into the slab. This will require
+integrating the LRU a bit more closely into the query storage so that, given an
+index, it can access and modify the LRU index data.
+
+### Data structure choices and requirements
+
+This section discusses the choice of crates and the actual requirements
+for each datastructure. 
+
+For `key_indices`, the [dashmap] crate appears to be the best concurrent hashmap
+available. It is more general than we need in that it supports multiple parallel
+writers, but we only ever have one, and it supports removal of keys and other
+such features. On the other hand, it is also less capable than we might like, in
+that unlike `IndexMap` it doesn't support indexing (and hence we must store the
+`Q::Key` in the `key_data` array).
+
+For `key_data`, shared-slab is used but it has several capabilities we don't
+require. We only have a single writer at a time (but that writer executes in
+parallel with readers). We don't remove entries and don't really require
+generational indexing.
+
+[dashmap]: https://crates.io/crates/dashmap
+
+## Alternatives and future work
+
+### Dropping cached values eagerly
+
+This RFC does not address the desire, expressed by maklad in the (unmerged)
+[better defaults RFC], to have queries whose cached values are automatically
+dropped when we enter the next revision. This could be readily accommodated as
+an extension to the hierarchy, though I'm unsure if this problem was sufficiently
+addressed by the need for LRU. 
+
+[better defaults RFC]: https://github.com/salsa-rs/salsa-rfcs/pull/4
+
+### Values that are not `Eq`
+
+This RFC does not address the problem that we cannot accommodate queries whose
+values do not implement `Eq` but which *are* cached. We used to permit this
+with volatile queries, but we no longer support those. Note that if we permitted
+values to be cached but only within one revision, this would address the same
+use cases as volatile (particularly combined with volatile reads).
+
+### Backdating with hashes
+
+Another way to handle "backdating" besides using `Eq` is to permit storing a
+cryptographic hash of the value instead of the value itself, which would permit
```

Comment:
Unimportant, but is there a better term than "cryptographic" here? We don't need hashing to be one-way, we just need the probability of collision to be negligible.